internal/repo/update.go

```diff
@@ -0,0 +1,153 @@
+)
+
+func UpdateRepo() error {
+	rel, err := getLatestRelease()
+	if err != nil {
+		return fmt.Errorf("failed to get latest release: %v", err)
+	}
+
+	if !canUpdate(rel) {
+		log.Printf("current version is already the latest version, no update \n")
+		return nil
+	}
+
+	//Filtering assets by GOOS and GOARCH
+	if tAsset = getTargetAsset(rel, false); tAsset == nil {
+		return fmt.Errorf("no target asset found for %s %s", constant.OS, constant.Arch)
+	}
+	if shaAsset = getTargetAsset(rel, true); shaAsset == nil {
+		return fmt.Errorf("no sha256 asset found for %s %s", constant.OS, constant.Arch)
+	}
+
+	//Download the new version nali and its sha256
+	data, err := download(ctx, tAsset.GetID())
+	if err != nil {
+		return fmt.Errorf("failed to download asset %v: %v", tAsset.GetID(), err)
+	}
+
+	vData, err := download(ctx, shaAsset.GetID())
+	if err != nil {
+		return fmt.Errorf("failed to download asset %v: %v", tAsset.GetID(), err)
+	}
+
+	// Verifying files with sha256
+	vHash := make([]byte, sha256.Size)
+	if _, err := hex.Decode(vHash, vData[:sha256.BlockSize]); err != nil {
+		return fmt.Errorf("failed to decode sha256 hash: %v", err)
+	}
+	if !validate(data, vHash) {
+		return fmt.Errorf("failed to validate asset %v, sha256 check failed", tAsset.GetID())
+	}
+
+	// Unzip and replace nali itself
+	exe, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("could not locate executable path: %v", err)
+	}
+
+	asset, err := decompress(bytes.NewReader(data), tAsset.GetName())
+	if err != nil {
+		return fmt.Errorf("error occurred while decompress: %v", err)
+	}
+
+	log.Printf("Updating %v to %v \n", exe, rel.GetTagName())
+	if err = update(asset, exe); err != nil {
+		return fmt.Errorf("update executable failed: %v", err)
+	}
+
+	log.Printf("Successfully updated to version %v \n", rel.GetTagName())
+	return nil
+}
+
+func canUpdate(rel *github.RepositoryRelease) bool {
+	if constant.Version != "unknown version" {
+		latest, _ := semver.NewVersion(rel.GetTagName())
+		cur, _ := semver.NewVersion(constant.Version)
+
+		return latest.GreaterThan(cur)
+	}
+	return true
+}
+
+func update(asset io.Reader, cmdPath string) error {
+	newBytes, err := io.ReadAll(asset)
+	if err != nil {
+		return err
+	}
+
+	// get the directory the executable exists in
+	updateDir := filepath.Dir(cmdPath)
+	filename := filepath.Base(cmdPath)
+
+	// Copy the contents of new binary to a new executable file
+	newPath := filepath.Join(updateDir, fmt.Sprintf(".%s.new", filename))
+	fp, err := os.OpenFile(newPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0755)
+
+	if err != nil {
+		fp.Close()
+		return fmt.Errorf("create the new executable file failed: %v", err)
+	}
+
+	if _, err = io.Copy(fp, bytes.NewReader(newBytes)); err != nil {
+		fp.Close()
+		return fmt.Errorf("copy the new executable file failed: %v", err)
+	}
+	// if we don't call fp.Close(), windows won't let us move the new executable
+	// because the file will still be "in use"
+	fp.Close()
```

Comment:
We should handle the error here, although the probability of the error here is extremely low, we need to make sure that we don't cause file corruption here, and if an error occurs we need to cancel the update process